### ISSUE TITLE
Cannot link Library and deploy just using "@truffle/contract"

### ISSUE THREAD

**Issue by elranu** [trufflesuite/truffle#3584]
- [x] I've asked for help in the [Truffle Gitter](http://gitter.im/Consensys/truffle) before filing this issue.

---------------------------

## Issue

Cannot link Library just using "@truffle/contract". If I use the JS truffle test suit works.
## Steps to Reproduce

The contract is deployed correctly with the following code on the JS truffle test:
```javascript
        const lib = await ProxyLib.deployed();
        ProxyAccountGuardedV2.link(ProxyLib, lib.address);
        const proxyContract = await ProxyAccountGuardedV2.new(owner, guardian);
```

But it does not work the same contracts with the following code just using "@truffle/contract",
        
```javascript
        var contract = require("@truffle/contract");

        const proxyLib = contract(proxyLibJson);
        proxyLib.setProvider(this.web3.currentProvider);
        
        await proxyLib.detectNetwork();
        const libInstance = await proxyLib.at(this.proxyLibAddress);
        
        this.proxyContract = contract(this.jsonProxyAccount);

        this.proxyContract.setProvider(this.web3.currentProvider);
        
        await this.proxyContract.detectNetwork();
        await this.proxyContract.link(libInstance);
        
        this.instanceProxy = await this.proxyContract.new(this.addressOwner, this.addressGuardian); <--- I get the error here

        console.log('Proxy address (owner): ' + this.instanceProxy.address);
```

## Expected Behavior

Link the contract and deploy.

## Actual Results
I get the following error:

```
TypeError: library_address.replace is not a function

 at processTicksAndRejections (internal/process/task_queues.js:93:5) {hijackedStack: 'TypeError: library_address.replace is not…ns (internal/process/task_queues.js:93:5)', stack: 'TypeError: library_address.replace is not a f…ctions (internal/process/task_queues.js:93:5)', message: 'library_address.replace is not a function'}
src/services/eth-did-service.ts:72
arg0:TypeError: library_address.replace is not a function
    at ProxyAccountGuarded.<anonymous> (C:\Dev\repos\did-tools\src\tools\proxy-account.ts:62:58)
    at step (C:\Dev\repos\did-tools\src\tools\proxy-account.ts:33:23)
    at Object.next (C:\Dev\repos\did-tools\src\tools\proxy-account.ts:14:53)
    at fulfilled (C:\Dev\repos\did-tools\src\tools\proxy-account.ts:5:58)
    at processTicksAndRejections (internal/process/task_queues.js:93:5) {hijackedStack: 'TypeError: library_address.replace is not…ns (internal/process/task_queues.js:93:5)', stack: 'TypeError: library_address.replace is not a f…ctions (internal/process/task_queues.js:93:5)', message: 'library_address.replace is not a function'}
```

## Environment

* Operating System: Win 10
* Ethereum client: Web3js
* Truffle version (`truffle version`): v5.1.55 (core: 5.1.55)
* node version (`node --version`): v14.15.1
* npm version (`npm --version`): Web3.js v1.2.9

**Comment by gnidan** [comment]
Thanks for opening this @elranu! After some discussion, this sounds like we just need to get to the bottom of it. Moving this to our backlog, but we'll let you know if we have follow-up questions. Appreciate your bringing this to our attention!

**Comment by elranu** [comment]
I will try to send you a repo with the error.

**Comment by elranu** [comment]
@gnidan here is a repo with metacoin to reproduce it. Hope it helps https://github.com/elranu/truffle-contract-bug

Thank you for helping!

**Comment by eggplantzzz** [comment]
Hey @elranu, thanks for the repro repo! I got down to the bottom of this this morning; it turns out that currently Truffle expects you to use a "contract type" (like `contract(MyJson)`) and not a "contract abstraction" (like `(await contract(<MyJson>)).deployed()`. I assume the above example will correctly create a new contract if you use `proxyLib` as an argument for `link`.

However, we probably should allow for the instance as an argument as well. It actually seems more appropriate to use the abstraction rather than the type as a type could have multiple deployments on the network.

**Comment by elranu** [comment]
Crap, I wasted lot of time. And when I did what you suggest I had another silly error I did not fix, so I did not realized. 

Thanks... I tested as you said and is correct. it works.

**Comment by eggplantzzz** [comment]
@elranu No problem! I should have a bug fix out in the next couple of days or so that will allow you to use the contract instances. That does seem more appropriate anyways.

**Comment by eggplantzzz** [comment]
This should be fixed in version `5.1.58`! Thanks for the report @elranu!
